cli/network.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2018 Huawei Corporation.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/kata-containers/agent/protocols/grpc"
+	vc "github.com/kata-containers/runtime/virtcontainers"
+	"github.com/urfave/cli"
+)
+
+var networkCLICommand = cli.Command{
+	Name:  "network",
+	Usage: "manage interfaces and routes for container",
+	Subcommands: []cli.Command{
+		addIfaceCommand,
+		delIfaceCommand,
+		listIfacesCommand,
+		updateRoutesCommand,
+		listRoutesCommand,
+	},
+	Action: func(context *cli.Context) error {
+		return cli.ShowSubcommandHelp(context)
+	},
+}
+
+var addIfaceCommand = cli.Command{
+	Name:      "add-iface",
+	Usage:     "add an interface to a container",
+	ArgsUsage: `add-iface <container-id> file or - for stdin`,
+	Flags:     []cli.Flag{},
+	Action: func(context *cli.Context) error {
+		if context.Args().Present() == false {
+			return fmt.Errorf("missing container ID, should provide one")
+		}
+		containerID := context.Args().First()
+		status, sandboxID, err := getExistingContainerInfo(containerID)
+		if err != nil {
+			return err
+		}
+		containerID = status.ID
+		// container MUST be running
+		if status.State.State != vc.StateRunning {
+			return fmt.Errorf("container %s is not running", containerID)
+		}
+
+		var (
+			inf   *grpc.Interface
+			f     *os.File
+			input = context.Args().Get(1)
+		)
+		if input == "-" {
+			f = os.Stdin
+		} else {
+			f, err = os.Open(input)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+		}
+		if err = json.NewDecoder(f).Decode(&inf); err != nil {
+			return err
+		}
+
+		resultingInf, err := vci.AddInterface(sandboxID, inf)
+		fmt.Println(resultingInf)
+		return err
+	},
+}
+
+var delIfaceCommand = cli.Command{
+	Name:      "del-iface",
+	Usage:     "delete an interface from a container",
+	ArgsUsage: `del-iface <container-id> file or - for stdin`,
+	Flags:     []cli.Flag{},
+	Action: func(context *cli.Context) error {
```

Comment:
This anonymous function is 99% identical to the one for `addIfaceCommand` - I'd create a generic `toggleIfaceCommand(add bool)` which will call `vci.AddInterface()` if `add == true`, else `vci.RemoveInterface()`.

Also the function in `updateRoutesCommand()` is extremely similar too if that could also be dealt with in a common way?

Reply:
Combine these three functions to one.